File: leetcode/1091. Shortest Path in Binary Matrix.py
from typing import List
from collections import deque
from collections import deque
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        n = len(grid)
        # 상 하 좌 우 왼
        dy = [-1, 1, 0, 0, -1, +1, -1, 1]
        dx = [0, 0, -1, 1, -1, -1, 1, 1]
        visited = [[0 for _ in range(n)] for _ in range(n)]
        if grid[0][0] == 1 or grid[n-1][n-1] == 1:
            return -1
        queue = deque([[0, 0, 1]]) #시작노드 push
        grid[0][0] = 1 #시작노드 방문처리
        while queue:
            y, x, cnt = queue.popleft()
            if y == n-1 and x == n-1: #도착한 경우
                return cnt
            for k in range(8): #8방향 탐색 시작
                ny = y + dy[k]
                nx = x + dx[k]
                if ny < 0 or ny >= n or nx < 0 or nx >= n:
                    continue
                if grid[ny][nx] == 0: #갈 수 있다면
                    queue.append([ny, nx, cnt + 1])
                    grid[ny][nx] = 1 #방문처리
        return -1






grid = [[0,1],[1,0]]

# 테스트 실행
solution = Solution()
result = solution.shortestPathBinaryMatrix(grid)
print(result) #4

File: 1091-shortest-path-in-binary-matrix/1091-shortest-path-in-binary-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {
    const n = grid.length;
    const queue = [[0, 0]];
    let head = 0;
    //방향 설정하기 (상, 하, 좌, 우, 왼쪽위, 왼쪽아래, 오른쪽위, 오른쪽아래)
    const dr = [-1, 1, 0, 0, -1,-1, 1, 1 ];
    const dc = [0, 0, -1, 1, 1, -1, -1, 1];
    const visited = Array.from({length : n}, () => Array(n).fill(-1));
    visited[0][0] = 1;
    if (grid[0][0] === 1 || grid[n-1][n-1] === 1) return -1;
    while (head < queue.length) {
        const [r, c] = queue[head++];
        for (let i = 0; i < 8; i++) {
            const nr = r + dr[i];
            const nc = c + dc[i];
            if (nr < 0 || nr >= n || nc < 0 || nc >= n) continue;
            if (visited[nr][nc] === -1 && grid[nr][nc] === 0) {
                queue.push([nr,nc])
                visited[nr][nc] = visited[r][c] + 1;
            }
        }
    }
    console.log(visited);
    if (visited[n-1][n-1] === 0) return -1;
    return visited[n-1][n-1];
};